src/algebra/module/projective.lean

```diff
@@ -0,0 +1,143 @@
+The actual definition of projective we use is that the natural R-module map
+from the free R-module on the type M down to M splits. This is more convenient
+than certain other definitions which involve quantifying over universes,
+and also universe-polymorphic (the ring and module can be in different universes).
+
+Everything works for semirings and semimodules except that apparently
+we don't have free semimodules, so here we stick to rings.
+
+`is_projective.of_lifting_property` is not universe polymorphic.
+
+## References
+
+https://en.wikipedia.org/wiki/Projective_module
+
+## Tags
+
+projective module
+
+-/
+
+universes u v
+
+/- The actual implementation we choose: `M` is projective if the natural surjection
+   from the free `R`-module on `M` to `M` splits. -/
+/-- An R-module is projective if it is a direct summand of a free module, or equivalently
+  if maps from the module lift along surjections. There are several other equivalent
+  definitions. -/
+def is_projective
+  (R : Type u) [semiring R] (M : Type v) [add_comm_monoid M] [semimodule R M] : Prop :=
+∃ s : M →ₗ[R] (M →₀ R), function.left_inverse (finsupp.total M M R id) s
+
+namespace is_projective
+
+section semiring
+
+variables {R : Type u} [semiring R] {M : Type v} [add_comm_monoid M] [semimodule R M]
+  {A : Type*} [add_comm_group A] [semimodule R A] {B : Type*} [add_comm_group B] [semimodule R B]
+
+/-- A projective R-module has the property that maps from it lift along surjections. -/
+theorem lifting_property (h : is_projective R M) (f : A →ₗ[R] B) (g : M →ₗ[R] B)
+  (hf : function.surjective f) : ∃ (h : M →ₗ[R] A), f.comp h = g :=
+begin
+  /-
+  Recall that `X →₀ R` is Lean's way of talking about the free `R`-module
+  on a type `X`. The universal property `finsupp.total` says that to a map
+  `X → A` from a type to an `R`-module, we get an associated R-module map
+  `(X →₀ R) →ₗ A`. Apply this to a (noncomputable) map `M → A` coming from the map
+  `M → B` and a random splitting of the surjection `A → B`, and we get
+  a map `φ : (M →₀ R) →ₗ A`.
+  -/
+  let φ : (M →₀ R) →ₗ[R] A := finsupp.total _ _ _ (λ m, function.surj_inv hf (g m)),
+  -- By projectivity we have a map `M →ₗ (M →₀ R)`;
+  cases h with s hs,
+  -- Compose to get `M →ₗ A`. This works.
+  use φ.comp s,
+  ext m,
+  conv_rhs {rw ← hs m},
+  simp [φ, finsupp.total_apply, function.surj_inv_eq hf],
+end
+
+/-- A module which satisfies the universal property is projective. Note that this result
+only has one universe variable. -/
+theorem of_lifting_property {R : Type u} [semiring R]
+  {M : Type u} [add_comm_monoid M] [semimodule R M]
+  -- If for all surjections of R-modules A →ₗ B, all maps M →ₗ B lift to M →ₗ A,
+  (huniv : ∀ {A B : Type u} [add_comm_monoid A] [add_comm_monoid B],
+    by exactI
+    ∀ [semimodule R A] [semimodule R B],
+    by exactI
+    ∀ (f : A →ₗ[R] B) (g : M →ₗ[R] B),
+  function.surjective f → ∃ (h : M →ₗ[R] A), f.comp h = g) :
+  -- then M is projective.
+  is_projective R M :=
```

Comment:
This is at least more universe polymorphic:
```suggestion
/-- A module which satisfies the universal property is projective. Note that the universe variables
in `huniv` are somewhat restricted. -/
theorem of_lifting_property {R : Type u} [semiring R]
  {M : Type v} [add_comm_monoid M] [semimodule R M]
  -- If for all surjections of R-modules A →ₗ B, all maps M →ₗ B lift to M →ₗ A,
  (huniv : ∀ {A : Type (max v u)} {B : Type v} [add_comm_monoid A] [add_comm_monoid B],
    by exactI
    ∀ [semimodule R A] [semimodule R B],
    by exactI
    ∀ (f : A →ₗ[R] B) (g : M →ₗ[R] B),
  function.surjective f → ∃ (h : M →ₗ[R] A), f.comp h = g) :
  -- then M is projective.
  is_projective R M :=
```

Reply:
Oh very nice! I tried some maxes but I couldn't figure out how to get it to work.